From 9d17ee7c8c5fbd72e35d2d2fa77274658e1c89f2 Mon Sep 17 00:00:00 2001
From: Alban Crequy <alban@endocode.com>
Date: Sun, 29 Mar 2015 14:51:23 +0200
Subject: [PATCH] nspawn: fallback on bind mount when mknod fails

Some systems abusively restrict mknod, even when the device node already
exists in /dev. In that case, try to fallback on bind mount.
---
 src/nspawn/nspawn.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/src/nspawn/nspawn.c b/src/nspawn/nspawn.c
index c851fdd..e361788 100644
--- a/src/nspawn/nspawn.c
+++ b/src/nspawn/nspawn.c
@@ -1449,8 +1449,15 @@ static int copy_devnodes(const char *dest) {
                                 return -r;
                         }
 
-                        if (mknod(to, st.st_mode, st.st_rdev) < 0)
-                                return log_error_errno(errno, "mknod(%s) failed: %m", to);
+                        if (mknod(to, st.st_mode, st.st_rdev) < 0) {
+                                if (errno != EPERM)
+                                        return log_error_errno(errno, "mknod(%s) failed: %m", to);
+
+                                /* Some systems abusively restrict mknod but
+                                 * allow bind mounts. */
+                                if (mount(from, to, "bind", MS_BIND, NULL) < 0)
+                                        return log_error_errno(errno, "both mknod and bind mount (%s) failed: %m", to);
+                        }
 
                         if (arg_userns && arg_uid_shift != UID_INVALID)
                                 if (lchown(to, arg_uid_shift, arg_uid_shift) < 0)
-- 
2.1.4

